wagtail/admin/compare.py

```diff
@@ -55,8 +57,155 @@ def htmldiff(self):
+    def is_deleted(self):
+        return bool(self.val_a) and not bool(self.val_b)
+
+    def has_changed(self):
+        return self.val_a != self.val_b
+
+
+class CharBlockComparison(BlockComparison):
+    def htmldiff(self):
+        return diff_text(
+            force_text(self.val_a),
+            force_text(self.val_b)
+        ).to_html()
+
+
+class RichTextBlockComparison(BlockComparison):
+    def htmldiff(self):
+        return diff_text(
+            BeautifulSoup(force_text(self.val_a), 'html5lib').getText(),
+            BeautifulSoup(force_text(self.val_b), 'html5lib').getText()
+        ).to_html()
+
+
+class StructBlockComparison(BlockComparison):
+    def htmldiff(self):
+        htmldiffs = {}
+        for name, block in self.block.child_blocks.items():
+            comparison_class = get_comparison_class_for_block(block)
+
+            val_a = self.val_a[name]
+            val_b = self.val_b[name]
+
+            htmldiffs[name] = comparison_class(block, val_a, val_b).htmldiff()
+
+        return self.block.render_basic(htmldiffs)
+
+
+class StreamBlockComparison(BlockComparison):
+    def get_block_comparisons(self):
+        a_blocks = list(self.val_a) or []
+        b_blocks = list(self.val_b) or []
+
+        a_blocks_by_id = {block.id: block for block in a_blocks}
+        b_blocks_by_id = {block.id: block for block in b_blocks}
+
+        deleted_ids = a_blocks_by_id.keys() - b_blocks_by_id.keys()
+
+        comparisons = []
+        for block in b_blocks:
+            comparison_class = get_comparison_class_for_block(block.block)
+
+            if block.id in a_blocks_by_id:
+                # Changed/existing block
+                comparisons.append(comparison_class(block.block, a_blocks_by_id[block.id].value, block.value))
+            else:
+                # New block
+                comparisons.append(comparison_class(block.block, None, block.value))
+
+        # Insert deleted blocks at the index where they used to be
+        deleted_block_indices = [(block, i) for i, block in enumerate(a_blocks) if block.id in deleted_ids]
+
+        for block, index in deleted_block_indices:
+            comparison_class = get_comparison_class_for_block(block.block)
+
+            # Insert the block back in where it was before it was deleted.
+            # Note: we need to account for new blocks when finding the position.
+            current_index = 0
+            block_inserted = False
+            for i, comparison in enumerate(comparisons):
+                if comparison.is_new():
+                    continue
+
+                if current_index == index:
+                    comparisons.insert(i, comparison_class(block.block, block.value, None))
+                    block_inserted = True
+                    break
+
+                current_index += 1
+
+            # Deleted block was from the end
+            if not block_inserted:
+                comparisons.append(comparison_class(block.block, block.value, None))
```

Comment:
Can avoid a bit of DRY by moving the definition of `comparison_class(block.block, block.value, None)` up to the top of the outer loop.